File: VirtualGrid/ServerSidePaging/ServerSidePaging/ServerSidePagingVirtualGrid.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using Telerik.WinControls.UI;

namespace ServerSidePaging
{
    #region PageIndexChangingEvent
    public partial class ServerSidePagingVirtualGrid : Telerik.WinControls.UI.RadForm
    {
        private VirtualGridRepository repository;
        private IList<OrderDataModel> data;

        public ServerSidePagingVirtualGrid()
        {
            InitializeComponent();

            this.repository = new VirtualGridRepository();
            this.radVirtualGrid1.RowCount = this.repository.Orders.Count();
            this.radVirtualGrid1.ColumnCount = this.repository.ColumnNames.Length;

            this.radVirtualGrid1.EnablePaging = true;
            this.radVirtualGrid1.AutoSizeColumnsMode = Telerik.WinControls.UI.VirtualGridAutoSizeColumnsMode.Fill;
            this.SelectData(this.radVirtualGrid1.PageIndex);

            this.radVirtualGrid1.CellValueNeeded += RadVirtualGrid1_CellValueNeeded;
            this.radVirtualGrid1.PageChanging += RadVirtualGrid1_PageChanging;
        }

        private void RadVirtualGrid1_PageChanging(object sender, VirtualGridPageChangingEventArgs e)
        {
            this.SelectData(e.NewIndex);
        }

        private void RadVirtualGrid1_CellValueNeeded(object sender, Telerik.WinControls.UI.VirtualGridCellValueNeededEventArgs e)
        {
            if (e.ColumnIndex < 0)
            {
                return;
            }

            if (e.RowIndex == RadVirtualGrid.HeaderRowIndex)
            {
                e.Value = this.repository.ColumnNames[e.ColumnIndex];
            }

            if (e.RowIndex >= 0 && e.RowIndex < this.data.Count * (this.radVirtualGrid1.PageIndex + 1))
            {
                int index = e.RowIndex - this.radVirtualGrid1.PageSize * this.radVirtualGrid1.PageIndex;
                e.Value = this.data[index][e.ColumnIndex];
            }
        }

        private void SelectData(int pageIndex)
        {
            this.data = this.repository.Orders.Skip(pageIndex * this.radVirtualGrid1.PageSize).Take(this.radVirtualGrid1.PageSize).ToList();
        }
    }
    #endregion
}

File: VirtualGrid/ServerSidePaging/ServerSidePagingVB/ServerSidePagingVirtualGrid.vb
Imports Telerik.WinControls.UI

#Region "PageIndexChangingEvent"
Public Class ServerSidePagingVirtualGrid

    Private repository As VirtualGridRepository
    Private data As IList(Of OrderDataModel)

    Public Sub New()
        InitializeComponent()
        Me.repository = New VirtualGridRepository()
        Me.RadVirtualGrid1.RowCount = Me.repository.Orders.Count()
        Me.RadVirtualGrid1.ColumnCount = Me.repository.ColumnNames.Length
        Me.RadVirtualGrid1.EnablePaging = True
        Me.RadVirtualGrid1.AutoSizeColumnsMode = Telerik.WinControls.UI.VirtualGridAutoSizeColumnsMode.Fill
        Me.SelectData(Me.RadVirtualGrid1.PageIndex)
        AddHandler Me.RadVirtualGrid1.CellValueNeeded, AddressOf RadVirtualGrid1_CellValueNeeded
        AddHandler Me.RadVirtualGrid1.PageChanging, AddressOf RadVirtualGrid1_PageChanging
    End Sub

    Private Sub RadVirtualGrid1_PageChanging(ByVal sender As Object, ByVal e As VirtualGridPageChangingEventArgs)
        Me.SelectData(e.NewIndex)
    End Sub

    Private Sub RadVirtualGrid1_CellValueNeeded(ByVal sender As Object, ByVal e As Telerik.WinControls.UI.VirtualGridCellValueNeededEventArgs)
        If e.ColumnIndex < 0 Then
            Return
        End If

        If e.RowIndex = RadVirtualGrid.HeaderRowIndex Then
            e.Value = Me.repository.ColumnNames(e.ColumnIndex)
        End If

        If e.RowIndex >= 0 AndAlso e.RowIndex < Me.data.Count * (Me.RadVirtualGrid1.PageIndex + 1) Then
            Dim index As Integer = e.RowIndex - Me.RadVirtualGrid1.PageSize * Me.RadVirtualGrid1.PageIndex
            e.Value = Me.data(index)(e.ColumnIndex)
        End If
    End Sub

    Private Sub SelectData(ByVal pageIndex As Integer)
        Me.data = Me.repository.Orders.Skip(pageIndex * Me.RadVirtualGrid1.PageSize).Take(Me.RadVirtualGrid1.PageSize).ToList()
    End Sub
End Class
#End Region
